File: app/src/main/java/com/backdoor/walcartandroidtest/View/Model/CategoryFragment/CategoryRootCatRecyclerAdapter.java
package com.backdoor.walcartandroidtest.View.Model.CategoryFragment;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.ColorStateList;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.backdoor.walcartandroidtest.R;

import java.util.List;


public class CategoryRootCatRecyclerAdapter extends RecyclerView.Adapter<CategoryRootCatRecyclerAdapter.MyViewHolder> {

    private List<CategoryRootCat_Data> contact;
    private Context mContext;
    private int catSelectedPosition = -1;

    public CategoryRootCatRecyclerAdapter(List<CategoryRootCat_Data> contacts, Context context) {
        this.contact = contacts;
        this.mContext = context;
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_category_root_cat, parent, false);
        return new MyViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull final MyViewHolder holder, final int position) {

        try {
            if (position == 0) {
                holder.view1.setVisibility(View.GONE);
            }
            if (position == getItemCount()) {
                holder.view2.setVisibility(View.GONE);
            }

            holder.parentLayout.setOnClickListener(view -> {
                Log.d("CategoryRootCat", "cat clicked");
                checkSelected();
                catSelectedPosition = holder.getAdapterPosition();
                holder.img.setImageTintList(ColorStateList.valueOf(mContext.getResources().getColor(R.color.catSelectedColor)));
                holder.tittle.setTextColor(mContext.getResources().getColor(R.color.catSelectedColor));
                holder.view1.setBackgroundColor(mContext.getResources().getColor(R.color.catSelectedColor));
                holder.view2.setBackgroundColor(mContext.getResources().getColor(R.color.catSelectedColor));

            });

            itemUpdateChangeBG(holder, position);

        } catch (Exception ex) {
            Log.d("CategoryRootCat", "error: " + ex.getMessage());
        }

    }

    private void itemUpdateChangeBG(MyViewHolder holder, int position) {
        if (position == catSelectedPosition) {
            holder.img.setImageTintList(ColorStateList.valueOf(mContext.getResources().getColor(R.color.catSelectedColor)));
            holder.tittle.setTextColor(mContext.getResources().getColor(R.color.catSelectedColor));
            holder.view1.setBackgroundColor(mContext.getResources().getColor(R.color.catSelectedColor));
            holder.view2.setBackgroundColor(mContext.getResources().getColor(R.color.catSelectedColor));
        } else {
            holder.img.setImageTintList(ColorStateList.valueOf(mContext.getResources().getColor(R.color.navNormalColor)));
            holder.tittle.setTextColor(mContext.getResources().getColor(R.color.secondaryTextColor));
            holder.view1.setBackgroundColor(mContext.getResources().getColor(R.color.secondaryTextColor));
            holder.view2.setBackgroundColor(mContext.getResources().getColor(R.color.secondaryTextColor));
        }
    }

    private void checkSelected() {
        if (!(catSelectedPosition < 0)) {
            notifyItemChanged(catSelectedPosition);
        }
    }

    @Override
    public int getItemCount() {
        return 10;
    }


    static class MyViewHolder extends RecyclerView.ViewHolder {

        TextView tittle;
        ImageView img;
        LinearLayout parentLayout;
        View view1, view2;

        MyViewHolder(View itemView) {
            super(itemView);
            parentLayout = itemView.findViewById(R.id.item_category_root_cat_layout);
            img = itemView.findViewById(R.id.item_category_root_cat_img);
            tittle = itemView.findViewById(R.id.item_category_root_cat_text);
            view1 = itemView.findViewById(R.id.item_category_root_cat_view1);
            view2 = itemView.findViewById(R.id.item_category_root_cat_view2);
        }
    }


    @SuppressLint("NotifyDataSetChanged")
    public void notifyChangeData(List<CategoryRootCat_Data> dataList) {
        contact = dataList;
        notifyDataSetChanged();
    }

}

File: app/src/main/java/com/backdoor/walcartandroidtest/View/Model/CategoryFragment/CategoryRootCatRecyclerAdapter.kt
package com.backdoor.walcartandroidtest.View.Model.CategoryFragment

import android.content.Context
import android.content.res.ColorStateList
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.backdoor.walcartandroidtest.R
import com.backdoor.walcartandroidtest.databinding.ItemCategoryRootCatBinding
import com.example.GetCategoriesListQuery
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso

class CategoryRootCatRecyclerAdapter(
    private val mContext: Context,
    private var contact: List<GetCategoriesListQuery.Category>?,
    private val mListener: OnItemClickListener?
) : RecyclerView.Adapter<CategoryRootCatRecyclerAdapter.MyViewHolder>() {
    private var catSelectedPosition = -1
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val binding =
            ItemCategoryRootCatBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return MyViewHolder(binding)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        try {

            holder.binding.itemCategoryRootCatText.text = contact?.get(position)?.enName ?: ""

            val imageUrl = contact?.get(position)?.image?.url

            if (imageUrl != null) {
                Picasso.get()
                    .load(imageUrl.toString())
                    .error(R.drawable.icons_no_image_50)
                    .into(holder.binding.itemCategoryRootCatImg, object : Callback {
                        override fun onSuccess() {
                            holder.binding.itemCategoryRootCatProgressBar.visibility = View.GONE
                            holder.binding.itemCategoryRootCatImg.visibility = View.VISIBLE
                        }

                        override fun onError(ex: java.lang.Exception) {
                            holder.binding.itemCategoryRootCatProgressBar.visibility = View.GONE
                            holder.binding.itemCategoryRootCatImg.visibility = View.VISIBLE
                            holder.binding.itemCategoryRootCatImg.setImageResource(R.drawable.icons_no_image_50)
                        }
                    })
            } else {
                holder.binding.itemCategoryRootCatProgressBar.visibility = View.GONE
                holder.binding.itemCategoryRootCatImg.visibility = View.VISIBLE
                holder.binding.itemCategoryRootCatImg.setImageResource(R.drawable.icons_no_image_50)
            }

            if (position == 0) {
                holder.binding.itemCategoryRootCatView1.visibility = View.GONE
            }
            if (position == itemCount) {
                holder.binding.itemCategoryRootCatView2.visibility = View.GONE
            }
            holder.binding.itemCategoryRootCatLayout.setOnClickListener {
                Log.d("CategoryRootCat", "cat clicked")
                mListener?.onItemClick(position)
                checkSelected()
                catSelectedPosition = holder.adapterPosition
                holder.binding.itemCategoryRootCatImg.imageTintList = ColorStateList.valueOf(
                    mContext.resources.getColor(R.color.catSelectedColor)
                )
                holder.binding.itemCategoryRootCatText.setTextColor(
                    mContext.resources.getColor(R.color.catSelectedColor)
                )
                holder.binding.itemCategoryRootCatView1.setBackgroundColor(
                    mContext.resources.getColor(R.color.catSelectedColor)
                )
                holder.binding.itemCategoryRootCatView2.setBackgroundColor(
                    mContext.resources.getColor(R.color.catSelectedColor)
                )
            }
            itemUpdateChangeBG(holder, position)
        } catch (ex: Exception) {
            Log.d("CategoryRootCat", "error: " + ex.message)
        }
    }

    private fun itemUpdateChangeBG(holder: MyViewHolder, position: Int) {
        if (position == catSelectedPosition) {
            holder.binding.itemCategoryRootCatImg.imageTintList =
                ColorStateList.valueOf(mContext.resources.getColor(R.color.catSelectedColor))
            holder.binding.itemCategoryRootCatText.setTextColor(mContext.resources.getColor(R.color.catSelectedColor))
            holder.binding.itemCategoryRootCatView1.setBackgroundColor(mContext.resources.getColor(R.color.catSelectedColor))
            holder.binding.itemCategoryRootCatView2.setBackgroundColor(mContext.resources.getColor(R.color.catSelectedColor))
        } else {
            holder.binding.itemCategoryRootCatImg.imageTintList =
                ColorStateList.valueOf(mContext.resources.getColor(R.color.navNormalColor))
            holder.binding.itemCategoryRootCatText.setTextColor(mContext.resources.getColor(R.color.secondaryTextColor))
            holder.binding.itemCategoryRootCatView1.setBackgroundColor(mContext.resources.getColor(R.color.secondaryTextColor))
            holder.binding.itemCategoryRootCatView2.setBackgroundColor(mContext.resources.getColor(R.color.secondaryTextColor))
        }
    }

    interface OnItemClickListener {
        fun onItemClick(position: Int)
    }

    private fun checkSelected() {
        if (catSelectedPosition >= 0) {
            notifyItemChanged(catSelectedPosition)
        }
    }

    override fun getItemCount(): Int {
        return if (contact == null) {
            0
        } else if (contact!!.isEmpty()) {
            0
        } else {
            contact!!.size
        }
    }

    class MyViewHolder(val binding: ItemCategoryRootCatBinding) :
        RecyclerView.ViewHolder(binding.root)

}
